src/3rd_party/topk.cuh

```diff
@@ -0,0 +1,302 @@
+                             T* outVals,
+                             const int beams_per_batch,
+                             const int k,
+                             bool descendingOrder) {
+
+  const int size = beams_per_batch * k * BLOCKS_PER_BEAM_; 
+  const int tid = threadIdx.x;
+  const int batch_id = blockIdx.x;
+  const T minimal = descendingOrder? -FpInfinity<T>::infinity() : FpInfinity<T>::infinity();;
+
+  typedef cub::BlockReduce<TopK<IndexType, T>, BLOCK_SIZE_> BlockReduce;
+  __shared__ typename BlockReduce::TempStorage temp_storage;
+  extern __shared__ char array[];
+  T *s_val = topk_tmp_val_buf + batch_id * size;
+  TopK<IndexType, T> *topks = (TopK<IndexType, T>*)(array);
+  
+  TopK<IndexType, T> partial;
+
+  for(int ite = 0; ite < k; ite++) {
+    partial.init(descendingOrder);
+    #pragma unroll
+    for(IndexType i = tid; i < size; i+= BLOCK_SIZE_) {
+      descendingOrder? partial.insertKeepMax(s_val[i], i) : partial.insertKeepMin(s_val[i], i);
+    }
+
+    TopK<IndexType, T> total = BlockReduce(temp_storage).Reduce(partial, descendingOrder? reduce_topk_max<IndexType, T>: reduce_topk_min<IndexType, T>);
+
+    if(tid == 0) {
+      topks[ite] = total;
+      s_val[total.p] = minimal;
+    }
+    __syncthreads();
+  }
+
+  for(int beam = tid; beam < k; beam += BLOCK_SIZE_) {
+    TopK<IndexType, T> beamOut; 
+    IndexType indexInTmpValRow = topks[beam].p;
+    beamOut.p = topk_tmp_id_buf[batch_id * size + indexInTmpValRow];
+    beamOut.u = topks[beam].u;
+    if(top) top[batch_id * k + beam] = beamOut;
+    if(outIndices) outIndices[batch_id * k + beam] = beamOut.p;
+    if(outVals) outVals[batch_id * k + beam] = beamOut.u;
+  }
+}
+
+#define CASE_K(K,BLOCK_SIZE_1_, BLOCK_SIZE_2_, BLOCKS_PER_BEAM_) \
+  case K: \
+    topk_stage_1<IndexType, T, BLOCK_SIZE_1_, BLOCKS_PER_BEAM_, getRowOffsets><<<batch_size * beams_per_batch * BLOCKS_PER_BEAM_, BLOCK_SIZE_1_, 0, stream>>>( \
+        log_probs, \
+        topk_tmp_id_buf, \
+        topk_tmp_val_buf, \
+        k, vocab_size, descendingOrder); \
+    topk_stage_2<IndexType, T, BLOCK_SIZE_2_, BLOCKS_PER_BEAM_><<<batch_size, BLOCK_SIZE_2_, K * sizeof(TopK<IndexType, T>), stream>>>( \
+        topk_tmp_id_buf, \
+        topk_tmp_val_buf, \
+        tops, \
+        outIndices, \
+        outVals, \
+        beams_per_batch, \
+        k, descendingOrder); \
+  break; \
+
+// The getRowOffsets template parameter is added so the topk implementation works with both nth_element.cu and the topk operator.
+// It is a template parameter since we know at compile time which version of topk we want to call. This flag can be removed whenever nth
+// element.cu is removed. When this flag is true, the indices returns are the offsets within the row. When the flag is false, the indices
+// returned are offset from the base pointer.
+template <typename IndexType, typename T, bool getRowOffsets=false>
+void topK_kernelLauncher(T* log_probs,
```

Comment:
In addition to the comment that mentions some detail information, please also add a function-level comment saying what this function is meant to do.